util/microservice-sdk/js/lib/kafka/Producer.js

```diff
@@ -0,0 +1,257 @@
+
+    if (this.isDeliveryReportEnabled()) {
+      // This will set a callback to be executed everytime a message is published.
+      const resolveOnDeliveryReportBind = Producer.resolveOnDeliveryReport.bind(this);
+      this.producer.on('delivery-report', resolveOnDeliveryReportBind);
+    }
+
+    this.producer.setPollInterval(this.config['producer.poll.internval.ms']);
+
+    logger.debug('... a Kafka producer was successfully created.', TAG);
+  }
+
+  /**
+  * Check if the delivery report is active
+  *
+  * To use the event 'delivery-report', you must set request.required.acks to 1 or -1 in topic
+  * configuration and dr_cb (or dr_msg_cb if you want the report to contain the message payload)
+  * to true in kafka options.
+  *
+  * @private
+  */
+  isDeliveryReportEnabled() {
+    return this.config.kafka.dr_cb || this.config.kafka.dr_msg_cb;
+  }
+
+  /**
+   * Connects the Producer to a Kafka cluster.
+   *
+   * This function will wait for 5 seconds in order to successfully connect to
+   * a broker instance. If there is any problem, then it will time out or
+   * node-rdkakfa will emit a event.error event, both of them triggering a
+   * rejection of the returned promise.
+   *
+   * @returns { Promise } A promise which will be resolved if this Producer is
+   * successfully connected to a Kafka broker, or reject if any errors occur.
+   */
+  connect() {
+    logger.info('Connecting the producer...', TAG);
+
+    return new Promise((resolve, reject) => {
+      // ready
+      // note: the ready function is called just once
+      this.producer.on('ready', () => {
+        logger.info('Producer is ready', TAG);
+        this.isReady = true;
+        return resolve();
+      });
+
+      // register handler for kafka events
+      // error
+      this.producer.on('event.error', (event) => {
+        logger.warn(`Kafka event.error: ${event}`, TAG);
+      });
+
+      // connect to kafka
+      this.producer.connect(undefined, (error) => {
+        if (error) {
+          logger.error(`Error on connect: ${error}`, TAG);
+          reject(error);
+        }
+      });
+    });
+  }
+
+  /**
+   * Produce a message to a Kafka broker.
+   *
+   * This function will create a message to be published to a Kafka broker. If
+   * message key is set, then all messages with the same key will use the same
+   * partition. If the message partition is set, then it will be used when
+   * producing the message to a Kafka broker.
+   *
+   *
+   * @param { string } topic The topic to which the message will be published
+   * @param { string } message The actual message
+   * @param { string } key (optional) -The key to be used to selected a partition.
+   * @param { number } partition (optional) - The particion to be used when publishing the
+   * message.
+   */
+  produce(topic, message, key = null, partition = null) {
+    if (this.isReady === false) {
+      logger.debug('It is not yet ready to produce.', TAG);
```

Comment:
if the client is still in the connection state, on initialization or on a reconnection, all the payloads will be ignored, this is the expected behavior?
Could we store these messages for a while (limited) then produce them?

Reply:
I think It's expected, since that `produce` must be called after `connect`.

Reply:
In fact, I don't know if there are other parameters
https://github.com/Blizzard/node-rdkafka#methods

Reply:
Ok, I understand.